File: client/app/src/main/java/com/rh/heji/widget/CardDecoration.java
package com.rh.heji.widget;

import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.util.TypedValue;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.lxj.xpopup.util.XPopupUtils;
import com.rh.heji.R;

import static com.rh.heji.ui.bill.adapter.BillNodeProviderKt.TYPE_TITLE;

public class CardDecoration extends RecyclerView.ItemDecoration {

    private void drawCardBackground(@NonNull Canvas c, @NonNull RecyclerView parent) {
        int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            View child = parent.getChildAt(i);// Item View
            RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
            int position = params.getViewAdapterPosition();
            if (parent.getAdapter().getItemViewType(position) == TYPE_TITLE) {
                child.setBackground(XPopupUtils.createDrawable(parent.getResources().getColor(R.color._xpopup_light_color), 15, 15, 0, 0));
            } else if (isLastInItemList(parent, position) || isLastItem(parent, position)) {
                child.setBackground(XPopupUtils.createDrawable(parent.getResources().getColor(R.color._xpopup_light_color), 0, 0, 15, 15));
            }
        }
    }

    @Override
    public void onDrawOver(@NonNull Canvas c, @NonNull RecyclerView
            parent, @NonNull RecyclerView.State state) {
        //super.onDrawOver(c, parent, state);
        drawCardBackground(c, parent);
    }

    @Override
    public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        Resources resources = parent.getContext().getResources();

        int padding = getPadding8dp(resources);

        RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) view.getLayoutParams();
        int position = params.getViewAdapterPosition();
        int viewType = parent.getAdapter().getItemViewType(position);

        if (viewType == TYPE_TITLE) {
            // header
            outRect.set(0, (int) (padding / 2), 0, 0);
        } else {
            if (isLastInItemList(parent, position) || isLastItem(parent, position)) {
                // last item before next header
                outRect.set(0, 0, 0, (int) (padding / 2));
            }
        }
//        outRect.inset((int) size16dp, 0);
        outRect.left = (int) padding;
        outRect.right = (int) padding;
        if (isLastItem(parent, position)) {//最后一个
            outRect.bottom = (int) padding / 2;
        }
    }

    private boolean isLastItem(RecyclerView parent, int position) {
        return position == parent.getAdapter().getItemCount() - 1;
    }

    private boolean isLastInItemList(RecyclerView parent, int position) {
        return parent.getAdapter().getItemViewType(position + 1) == TYPE_TITLE;
    }

    private int getPadding8dp(Resources resources) {
        float size16dp = 8f;
        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, size16dp, resources.getDisplayMetrics());
    }
}

File: client/app/src/main/java/com/rh/heji/widget/CardDecoration.kt
package com.rh.heji.widget

import android.content.res.Resources
import android.graphics.Canvas
import android.graphics.Rect
import android.util.TypedValue
import android.view.View
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import com.lxj.xpopup.util.XPopupUtils
import com.rh.heji.R
import com.rh.heji.ui.bill.adapter.TYPE_TITLE

class CardDecoration(val padding: Int = 8) : ItemDecoration() {
    private fun drawCardBackground(c: Canvas, parent: RecyclerView) {
        val childCount = parent.childCount
        for (i in 0 until childCount) {
            val child = parent.getChildAt(i) // Item View
            val params = child.layoutParams as RecyclerView.LayoutParams
            val position = params.viewAdapterPosition
            if (parent.adapter!!.getItemViewType(position) == TYPE_TITLE) {
                child.background = XPopupUtils.createDrawable(ContextCompat.getColor(parent.context,R.color._xpopup_light_color), 15f, 15f, 0f, 0f)
            } else if (isLastInItemList(parent, position) || isLastItem(parent, position)) {
                child.background = XPopupUtils.createDrawable(ContextCompat.getColor(parent.context,R.color._xpopup_light_color), 0f, 0f, 15f, 15f)
            }
        }
    }

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        //super.onDrawOver(c, parent, state);
        drawCardBackground(c, parent)
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        val resources = parent.context.resources
        val padding = getPadding(resources)
        val params = view.layoutParams as RecyclerView.LayoutParams
        val position = params.viewAdapterPosition
        val viewType = parent.adapter!!.getItemViewType(position)
        if (viewType == TYPE_TITLE) {
            // header
            outRect[0, (padding / 2), 0] = 0
        } else {
            if (isLastInItemList(parent, position) || isLastItem(parent, position)) {
                // last item before next header
                outRect[0, 0, 0] = (padding / 2)
            }
        }
        //        outRect.inset((int) size16dp, 0);
        outRect.left = padding
        outRect.right = padding
        if (isLastItem(parent, position)) { //最后一个
            outRect.bottom = padding / 2
        }
    }

    private fun isLastItem(parent: RecyclerView, position: Int): Boolean {
        return position == parent.adapter!!.itemCount - 1
    }

    private fun isLastInItemList(parent: RecyclerView, position: Int): Boolean {
        return parent.adapter!!.getItemViewType(position + 1) == TYPE_TITLE
    }

    private fun getPadding(resources: Resources): Int {
        val size16dp = padding * 1f
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, size16dp, resources.displayMetrics).toInt()
    }
}